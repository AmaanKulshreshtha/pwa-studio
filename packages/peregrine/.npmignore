.eslintrc.js
.storybook
/{dist,src}/**/__{docs,helpers,mocks,stories,tests}__/**
/scripts
