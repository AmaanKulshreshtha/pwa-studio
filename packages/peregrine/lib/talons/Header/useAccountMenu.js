import { useState, useCallback, useEffect } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useApolloClient, useMutation } from '@apollo/react-hooks';

import { useUserContext } from '@magento/peregrine/lib/context/user';
import { clearCartDataFromCache } from '@magento/peregrine/lib/Apollo/clearCartDataFromCache';
import { clearCustomerDataFromCache } from '@magento/peregrine/lib/Apollo/clearCustomerDataFromCache';

/**
 * The useAccountMenu talon complements the AccountMenu component.
 *
 * @returns {Object}    talonProps
 * @returns {Boolean}   talonProps.accountMenuIsOpen - Whether the menu that this trigger toggles is open or not.
 * @returns {Function} talonProps.setAccountMenuIsOpen - Function to set the value of talonProps.accountMenuIsOpen.
 */
export const useAccountMenu = props => {
    const { mutations, accountMenuIsOpen, setAccountMenuIsOpen } = props;
    const { signOut: signOutMutation } = mutations;

    const [view, setView] = useState('SIGNIN');
    const [username, setUsername] = useState('');

    const apolloClient = useApolloClient();
    const history = useHistory();
    const location = useLocation();
    const [revokeToken] = useMutation(signOutMutation);
    const [{ isSignedIn: isUserSignedIn }, { signOut }] = useUserContext();

    const handleSignOut = useCallback(async () => {
        setView('SIGNIN');
        setAccountMenuIsOpen(false);

        // Delete cart/user data from the redux store.
        await signOut({ revokeToken });
        await clearCartDataFromCache(apolloClient);
        await clearCustomerDataFromCache(apolloClient);

        // Refresh the page as a way to say "re-initialize". An alternative
        // would be to call apolloClient.resetStore() but that would require
        // a large refactor.
        history.go(0);
    }, [apolloClient, history, revokeToken, setAccountMenuIsOpen, signOut]);

    const handleForgotPassword = useCallback(() => {
        setView('FORGOT_PASSWORD');
    }, []);

    const handleCreateAccount = useCallback(() => {
        setView('CREATE_ACCOUNT');
    }, []);

    // Close the Account Menu on page change.
    useEffect(() => {
        setAccountMenuIsOpen(false);
    }, [location.pathname, setAccountMenuIsOpen]);

    // Update view based on user status everytime accountMenuIsOpen has changed.
    useEffect(() => {
        if (isUserSignedIn) {
            setView('ACCOUNT');
        } else {
            setView('SIGNIN');
        }
    }, [accountMenuIsOpen, isUserSignedIn]);

    return {
        view,
        username,
        isUserSignedIn,
        handleSignOut,
        handleForgotPassword,
        handleCreateAccount,
        updateUsername: setUsername
    };
};
