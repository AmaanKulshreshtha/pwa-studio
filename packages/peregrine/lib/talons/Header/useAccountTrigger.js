import { useState, useCallback, useEffect } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { useApolloClient, useMutation } from '@apollo/react-hooks';

import { useUserContext } from '@magento/peregrine/lib/context/user';
import { useDropdown } from '@magento/peregrine/lib/hooks/useDropdown';
import { clearCartDataFromCache } from '@magento/peregrine/lib/Apollo/clearCartDataFromCache';
import { clearCustomerDataFromCache } from '@magento/peregrine/lib/Apollo/clearCustomerDataFromCache';

/**
 * The useAccountTrigger talon complements the AccountTrigger component.
 *
 * @returns {Object}    talonProps
 * @returns {Boolean}   talonProps.accountMenuIsOpen - Whether the menu that this trigger toggles is open or not.
 * @returns {Ref}       talonProps.accountMenuRef - A React ref to the menu that this trigger toggles.
 * @returns {Ref}       talonProps.accountMenuTriggerRef - A React ref to the trigger element itself.
 * @returns {Function}  talonProps.handleSignOut - A function for handling sign out.
 * @returns {Function}  talonProps.handleTriggerClick - A function for handling when the trigger is clicked.
 * @returns {Boolean}   talonProps.isUserSignedIn - Whether there is a user currently signed in or not.
 */
export const useAccountTrigger = props => {
    const { mutations, VIEWS } = props;
    const { signOut: signOutMutation } = mutations;

    const [view, setView] = useState(VIEWS.SIGNIN);
    const [username, setUsername] = useState('');

    const {
        elementRef: accountMenuRef,
        expanded: accountMenuIsOpen,
        setExpanded: setAccountMenuIsOpen,
        triggerRef: accountMenuTriggerRef
    } = useDropdown();

    const apolloClient = useApolloClient();
    const history = useHistory();
    const location = useLocation();
    const [revokeToken] = useMutation(signOutMutation);
    const [{ isSignedIn: isUserSignedIn }, { signOut }] = useUserContext();

    const handleSignOut = useCallback(async () => {
        setView(VIEWS.SIGNIN);
        setAccountMenuIsOpen(false);

        // Delete cart/user data from the redux store.
        await signOut({ revokeToken });
        await clearCartDataFromCache(apolloClient);
        await clearCustomerDataFromCache(apolloClient);

        // Refresh the page as a way to say "re-initialize". An alternative
        // would be to call apolloClient.resetStore() but that would require
        // a large refactor.
        history.go(0);
    }, [
        apolloClient,
        history,
        revokeToken,
        setAccountMenuIsOpen,
        signOut,
        VIEWS.SIGNIN
    ]);

    const handleTriggerClick = useCallback(() => {
        // Toggle the Account Menu.
        setAccountMenuIsOpen(isOpen => !isOpen);
        /**
         * When the dropdown closes, the view has to
         * reset to SINGIN view if the user is not
         * signed in.
         */
        if (!isUserSignedIn) {
            setView(VIEWS.SIGNIN);
        }
    }, [setAccountMenuIsOpen, isUserSignedIn, VIEWS]);

    const handleForgotPassword = useCallback(() => {
        setView(VIEWS.FORGOT_PASSWORD);
    }, [VIEWS.FORGOT_PASSWORD]);

    const handleCreateAccount = useCallback(() => {
        setView(VIEWS.CREATE_ACCOUNT);
    }, [VIEWS.CREATE_ACCOUNT]);

    // Close the Account Menu on page change.
    useEffect(() => {
        setAccountMenuIsOpen(false);
    }, [location.pathname, setAccountMenuIsOpen]);

    // Set view to account if the user is signed in and view is not ACCOUNT
    useEffect(() => {
        if (isUserSignedIn && view !== VIEWS.ACCOUNT) {
            setView(VIEWS.ACCOUNT);
        }
    }, [isUserSignedIn, VIEWS, view]);

    return {
        accountMenuIsOpen,
        accountMenuRef,
        accountMenuTriggerRef,
        handleSignOut,
        handleTriggerClick,
        isUserSignedIn,
        handleForgotPassword,
        handleCreateAccount,
        view,
        username,
        setUsername
    };
};
