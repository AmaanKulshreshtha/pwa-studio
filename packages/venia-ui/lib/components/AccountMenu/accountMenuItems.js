import React, { Fragment } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { func, shape, string } from 'prop-types';

import { Link } from '@magento/venia-drivers';
import { mergeClasses } from '@magento/venia-ui/lib/classify';

import defaultClasses from './accountMenuItems.css';

const AccountMenuItems = props => {
    const { handleSignOut } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    const intl = useIntl();
    const MENU_ITEMS = [
        { name: intl.formatMessage({ id: 'Order History' }), url: '' },
        {
            name: intl.formatMessage({ id: 'Store Credit & Gift Cards' }),
            url: ''
        },
        { name: intl.formatMessage({ id: 'Favorites Lists' }), url: '' },
        { name: intl.formatMessage({ id: 'Address Book' }), url: '' },
        { name: intl.formatMessage({ id: 'Saved Payments' }), url: '' },
        { name: intl.formatMessage({ id: 'Communications' }), url: '' },
        { name: intl.formatMessage({ id: 'Account Information' }), url: '' }
    ];
    const menuItems = MENU_ITEMS.map(item => {
        return (
            <Link className={classes.link} to={item.url} key={item.name}>
                {item.name}
            </Link>
        );
    });

    return (
        <Fragment>
            {menuItems}
            <button
                className={classes.signOut}
                onClick={handleSignOut}
                type="button"
            >
                <FormattedMessage id={'Sign Out'} />
            </button>
        </Fragment>
    );
};

export default AccountMenuItems;

AccountMenuItems.propTypes = {
    classes: shape({
        link: string,
        signOut: string
    }),
    handleSignOut: func
};
