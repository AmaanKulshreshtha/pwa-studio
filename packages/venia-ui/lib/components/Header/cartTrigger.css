.trigger {
    composes: root from '../clickable.css';
    height: 3rem;
    min-width: 3rem;
    position: relative;
    z-index: 1;
    display: flex;
    align-content: center;
    justify-content: center;
}

.link {
    composes: trigger;
    
    display: none;
}

.background {
    align-items: center;
    display: grid;
    /* The full height of the site header. */
    height: 5rem;
}

.background_open {
    composes: background;

    background-color: rgb(var(--venia-global-color-background));
    /* Sit on top of the mask created by the MiniCart. */
    z-index: 4;
}

.counter {
    top: 0.375rem;
    position: absolute;
    left: 1.5rem;
    font-weight: 600;
    box-sizing: border-box;
    padding: 0 0.25rem;
    background-color: rgb(var(--venia-brand-color-1-700));
    border-radius: 100px;
    color: rgb(var(--venia-global-color-background));
    line-height: 1.25rem;
    min-width: 1.25rem;
    font-size: 0.8125rem;
}

/*
 * Mobile-specific styles.
 */
@media (max-width: 960px) {
    .trigger {
        display: none;
    }

    .link {
        display: block;
    }
}
