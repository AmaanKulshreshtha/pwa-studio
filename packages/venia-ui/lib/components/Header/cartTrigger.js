import React, { Fragment } from 'react';
import { shape, string } from 'prop-types';
import { ShoppingCart as ShoppingCartIcon } from 'react-feather';

import { useCartTrigger } from '@magento/peregrine/lib/talons/Header/useCartTrigger';

import { mergeClasses } from '../../classify';
import CREATE_CART_MUTATION from '../../queries/createCart.graphql';
import GET_CART_DETAILS_QUERY from '../../queries/getCartDetails.graphql';
import Icon from '../Icon';
import ShoppingBag from '../ShoppingBag';
import defaultClasses from './cartTrigger.css';
import { GET_ITEM_COUNT_QUERY } from './cartTrigger.gql';

const shouldUseShoppingBag = process.env.APP_USE_SHOPPING_BAG;
console.log('shouldUseShoppingBag', shouldUseShoppingBag);

const CartTrigger = props => {
    const {
        handleClick,
        handleDesktopClick,
        handleMobileClick,
        itemCount,
        shoppingBagIsOpen,
        setShoppingBagIsOpen
    } = useCartTrigger({
        mutations: {
            createCartMutation: CREATE_CART_MUTATION
        },
        queries: {
            getCartDetailsQuery: GET_CART_DETAILS_QUERY,
            getItemCountQuery: GET_ITEM_COUNT_QUERY
        }
    });

    const classes = mergeClasses(defaultClasses, props.classes);
    const backgroundClass = shoppingBagIsOpen ? classes.background_open : classes.background;
    const isFilled = itemCount > 0;
    const iconClass = isFilled ? classes.icon_filled : classes.icon_empty;
    const iconClasses = { root: iconClass };
    const buttonAriaLabel = `Toggle mini cart. You have ${itemCount} items in your cart.`;

    const itemCounter = itemCount ? (
        <span className={classes.counter}>{itemCount}</span>
    ) : null;

    // Because this button behaves differently on desktop and mobile
    // we render two buttons that differ only in their click handler
    // and control which one displays via CSS.
    const cartTrigger = shouldUseShoppingBag ? (
        <Fragment>
            <div className={backgroundClass}>
                <button
                    aria-label={buttonAriaLabel}
                    className={classes.root_desktop}
                    onClick={handleDesktopClick}
                >
                    <Icon classes={iconClasses} src={ShoppingCartIcon} />
                    {itemCounter}
                </button>
            </div>
            <button
                aria-label={buttonAriaLabel}
                className={classes.root_mobile}
                onClick={handleMobileClick}
            >
                <Icon classes={iconClasses} src={ShoppingCartIcon} />
                {itemCounter}
            </button>
            <ShoppingBag isOpen={shoppingBagIsOpen} setIsOpen={setShoppingBagIsOpen} />
        </Fragment>
    ) : (
        <button
            aria-label={buttonAriaLabel}
            className={classes.root}
            onClick={handleClick}
        >
            <Icon classes={iconClasses} src={ShoppingCartIcon} />
            {itemCounter}
        </button>
    );
    
    return cartTrigger;
};

export default CartTrigger;

CartTrigger.propTypes = {
    classes: shape({
        icon_empty: string,
        icon_filled: string,
        root: string
    })
};
