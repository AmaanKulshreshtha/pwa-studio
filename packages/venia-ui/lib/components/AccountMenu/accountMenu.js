import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import { mergeClasses } from '@magento/venia-ui/lib/classify';

import SignIn from '../SignIn/signIn';
import AccountMenuItems from './accountMenuItems';
import { VIEWS } from '../Header/accountTrigger';
import defaultClasses from './accountMenu.css';

const AccountMenu = React.forwardRef((props, ref) => {
    const {
        handleSignOut,
        isOpen,
        view,
        handleForgotPassword,
        handleCreateAccount,
        updateUsername
    } = props;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;

    let dropdownContents = null;

    switch (view) {
        case VIEWS.ACCOUNT: {
            dropdownContents = (
                <AccountMenuItems handleSignOut={handleSignOut} />
            );

            break;
        }
        case VIEWS.FORGOT_PASSWORD: {
            dropdownContents = (
                <div className={classes.forgotPassword}>
                    To be handled in PWA-77
                </div>
            );

            break;
        }
        case VIEWS.CREATE_ACCOUNT: {
            dropdownContents = (
                <div className={classes.createAccount}>
                    To be handled in PWA-804
                </div>
            );

            break;
        }
        case VIEWS.SIGNIN:
        default: {
            dropdownContents = (
                <SignIn
                    setDefaultUsername={updateUsername}
                    showCreateAccount={handleCreateAccount}
                    showForgotPassword={handleForgotPassword}
                />
            );

            break;
        }
    }

    return (
        <aside className={rootClass} ref={ref}>
            {dropdownContents}
        </aside>
    );
});

export default AccountMenu;

AccountMenu.propTypes = {
    classes: shape({
        root: string,
        root_open: string,
        link: string
    }),
    handleSignOut: func,
    isOpen: bool,
    isUserSignedIn: bool,
    view: string,
    updateUsername: func.isRequired,
    handleCreateAccount: func.isRequired,
    handleForgotPassword: func.isRequired
};
