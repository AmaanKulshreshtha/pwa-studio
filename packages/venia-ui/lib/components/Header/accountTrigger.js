import React, { Fragment } from 'react';
import { shape, string } from 'prop-types';

import { useAccountTrigger } from '@magento/peregrine/lib/talons/Header/useAccountTrigger';
import { mergeClasses } from '@magento/venia-ui/lib/classify';

import AccountChip from '../AccountChip';
import AccountMenu from '../AccountMenu';

import SIGN_OUT_MUTATION from '../../queries/signOut.graphql';

import defaultClasses from './accountTrigger.css';

export const VIEWS = {
    SIGNIN: 'SIGNIN',
    FORGOT_PASSWORD: 'FORGOT_PASSWORD',
    CREATE_ACCOUNT: 'CREATE_ACCOUNT',
    ACCOUNT: 'ACCOUNT'
};

/**
 * The AccountTrigger component is the call to action in the site header
 * that toggles the AccountMenu dropdown.
 *
 * @param {Object} props
 * @param {Object} props.classes - CSS classes to override element styles.
 */
const AccountTrigger = props => {
    const talonProps = useAccountTrigger({
        mutations: { signOut: SIGN_OUT_MUTATION },
        VIEWS
    });
    const {
        accountMenuIsOpen,
        accountMenuRef,
        accountMenuTriggerRef,
        handleSignOut,
        handleTriggerClick,
        handleForgotPassword,
        handleCreateAccount,
        view,
        username,
        setUsername
    } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClassName = accountMenuIsOpen ? classes.root_open : classes.root;

    return (
        <Fragment>
            <div className={rootClassName} ref={accountMenuTriggerRef}>
                <button
                    aria-label={'Toggle My Account Menu'}
                    className={classes.trigger}
                    onClick={handleTriggerClick}
                >
                    <AccountChip
                        fallbackText={'Sign In'}
                        shouldIndicateLoading={true}
                    />
                </button>
            </div>
            <AccountMenu
                ref={accountMenuRef}
                view={view}
                username={username}
                isOpen={accountMenuIsOpen}
                onSignOut={handleSignOut}
                onForgotPassword={handleForgotPassword}
                onCreateAccount={handleCreateAccount}
                updateUsername={setUsername}
            />
        </Fragment>
    );
};

export default AccountTrigger;

AccountTrigger.propTypes = {
    classes: shape({
        root: string,
        root_open: string,
        trigger: string
    })
};
