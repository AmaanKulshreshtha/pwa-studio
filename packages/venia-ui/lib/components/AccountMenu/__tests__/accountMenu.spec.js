import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import AccountMenu from '../accountMenu';
import { VIEWS } from '../../Header/accountTrigger';

jest.mock('../accountMenuItems', () => 'AccountMenuItems');
jest.mock('../../SignIn/signIn', () => 'Singin Component');
jest.mock('../../ForgotPassword', () => 'Forgot Password Component');

const defaultProps = {
    handleSignOut: jest.fn().mockName('handleSignOut'),
    isOpen: true,
    isUserSignedIn: true,
    view: VIEWS.ACCOUNT,
    username: 'goosey_goose'
};

test('it renders AccountMenuItems when the user is signed in', () => {
    // Act.
    const instance = createTestInstance(<AccountMenu {...defaultProps} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('it renders SignIn component when the view is SIGNIN', () => {
    const props = {
        ...defaultProps,
        view: VIEWS.SIGNIN
    };

    // Act.
    const instance = createTestInstance(<AccountMenu {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('it renders forgot password component when the view is FORGOT_PASSWORD', () => {
    const props = {
        ...defaultProps,
        view: VIEWS.FORGOT_PASSWORD
    };

    // Act.
    const instance = createTestInstance(<AccountMenu {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('it renders forgot password component when the view is CREATE_ACCOUNT', () => {
    const props = {
        ...defaultProps,
        view: VIEWS.CREATE_ACCOUNT
    };

    // Act.
    const instance = createTestInstance(<AccountMenu {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});
