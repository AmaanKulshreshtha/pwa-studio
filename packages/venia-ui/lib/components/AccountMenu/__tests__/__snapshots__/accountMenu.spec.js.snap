// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders AccountMenuItems when the user is signed in 1`] = `
<aside>
  <AccountMenuItems
    handleSignOut={[MockFunction handleSignOut]}
  />
</aside>
`;

exports[`it renders SignIn component when the view is SIGNIN 1`] = `
<aside>
  <Singin Component />
</aside>
`;

exports[`it renders forgot password component when the view is CREATE_ACCOUNT 1`] = `
<aside>
  <div>
    To be handled in PWA-804
  </div>
</aside>
`;

exports[`it renders forgot password component when the view is FORGOT_PASSWORD 1`] = `
<aside>
  <div>
    To be handled in PWA-77
  </div>
</aside>
`;
